From e95487ead6134006f91d17178744a5f46adf567c Mon Sep 17 00:00:00 2001
From: Brennan Lamoreaux <brennan.lamoreaux@broadcom.com>
Date: Mon, 30 Jun 2025 18:01:30 +0000
Subject: [PATCH] FIPS: Mark structure field differences between kernel and
 canister

Use the newly added plugin attribute to mark the intentional type differences
between structure fields. These will be allowed to pass the matching
type check in the matching GCC plugin.

Signed-off-by: Brennan Lamoreaux <brennan.lamoreaux@broadcom.com>

[Alexey: Make it compatible with canister_build]
Signed-off-by: Alexey Makhalov <alexey.makhalov@broadcom.com>
---
 include/linux/plugin-attr.h | 9 +++++++++
 include/linux/swait.h       | 7 ++++++-
 2 files changed, 15 insertions(+), 1 deletions(-)
 create mode 100644 include/linux/plugin-attr.h

diff --git a/include/linux/plugin-attr.h b/include/linux/plugin-attr.h
new file mode 100644
index 000000000..386a90eac
--- /dev/null
+++ b/include/linux/plugin-attr.h
@@ -0,0 +1,9 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+
+/* This attribute is only registered by the match_canister_structs plugin */
+#ifdef MATCH_CANISTER_STRUCTS
+#define CANISTER_MISMATCH(KERN_TYPE, CANISTER_TYPE) \
+	__attribute__((canister_mismatch(KERN_TYPE, CANISTER_TYPE)))
+#else
+#define CANISTER_MISMATCH(KERN_TYPE, CANISTER_TYPE)
+#endif
diff --git a/include/linux/swait.h b/include/linux/swait.h
index d32441948..8e2d3010d 100644
--- a/include/linux/swait.h
+++ b/include/linux/swait.h
@@ -7,6 +7,7 @@
 #include <linux/spinlock.h>
 #include <linux/wait.h>
 #include <asm/current.h>
+#include <linux/plugin-attr.h>

 /*
  * Simple waitqueues are semantically very different to regular wait queues
@@ -41,7 +42,11 @@
 struct task_struct;

 struct swait_queue_head {
-	raw_spinlock_t		lock;
+#ifdef FIPS_CANISTER
+	unsigned char		lock[sizeof(raw_spinlock_t)];
+#else
+	raw_spinlock_t		lock CANISTER_MISMATCH("raw_spinlock_t", "unsigned char[4]");
+#endif
 	struct list_head	task_list;
 };

--
2.40.4

